instrumentation/net/http/otelhttpconv/DESIGN.md

```diff
@@ -0,0 +1,239 @@
+such as the HTTP route.
+
+When the data those options provide is not specified, the related span
+attributes will not be set.
+
+#### Request and Response wrappers
+
+The request and response wrappers implementations will be provided by the
+[internal/request](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/internal/request)
+templates package.
+
+We may provide a new package, or actual implementations within the
+`otelhttpconv` package later on. But doing so is not covered by this design
+document at the moment.
+
+Therefore external implementations will currently have to implement this logic
+themselves.
+
+### Implementations
+
+We will provide one official implementation of the described interfaces.
+This implementation will have the following requirements:
+
+* Support for the latest semantic conventions only.
+* Support for stable semantic conventions only.
+
+We may provide additional implementations later on such as:
+
+* An implementation that serves as a proxy to allow combining multiple implementations together.
+* An implementation that covers unstable semantic conventions.
+
+#### Example implementation
+
+The implementation example here is kept simple for the purpose of
+understandability.
+
+The following code sample provides a simple `http.Handler` that implements the
+provided interface to instrument HTTP applications.
+
+Because both the client and server interface are very similar, a client
+implementation would be similar too.
+
+```golang
+type middleware struct {
+	operation string
+
+	tracer      trace.Tracer
+	propagators propagation.TextMapPropagator
+	meter       metric.Meter
+	httpconv    otelhttpconv.Server
+}
+
+// NewMiddleware returns a tracing and metrics instrumentation middleware.
+// The handler returned by the middleware wraps a handler
+// in a span named after the operation and enriches it with metrics.
+func NewMiddleware(operation string) func(http.Handler) http.Handler {
+	m := middleware{
+		operation:   operation,
+		tracer:      otel.Tracer("http"),
+		propagators: otel.GetTextMapPropagator(),
+		meter:       otel.Meter("http"),
+		httpconv:    otelhttpconv.NewHTTPConv(otel.Tracer("httpconv"), otel.Meter("httpconv")),
+	}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			m.serveHTTP(w, r, next)
+		})
+	}
+}
+
+func (m *middleware) serveHTTP(w http.ResponseWriter, r *http.Request, next http.Handler) {
+	ctx := m.propagators.Extract(r.Context(), propagation.HeaderCarrier(r.Header))
+
+	// We keep creating the span here, as that is something we want to do before
+	// the middleware stack is run
+	ctx, span := m.tracer.Start(ctx, fmt.Sprintf("%s %s", r.Method, r.Pattern))
+	defer span.End()
+
+	// NewResponseWrapper wraps additional data into the http.ResponseWriter,
+	// such as duration, status code and quantity of bytes read.
+	rww := otelhttpconv.NewResponseWrapper(w)
```

Comment:
Isn't this coupled to `otelhttpconv.NewHTTPConv`?
Are we sure that it would work with other custom semantic conventions?

Reply:
So, even with structs for the wrappers, I think this design makes sense.
Building a response wrapper from an `HTTPConv` would be strange, as then each implementations would need to know how to build a wrapper as well.